File: Army(c++)/Warriors/Warlock.cpp
#include "Warlock.hpp"
#include "../State/State.hpp"

Warlock::Warlock(const std::string& name,  int hp, int mana, int dmg) 
    : SpellCaster(name, hp, mana, dmg, new SoldierState(this)), demonCost(40)
    , warrior(WarriorID::WARLOCK), magician(MagicianID::SORCERER) {}

Warlock::~Warlock() {}

WarriorID Warlock::getWarriorID() const {
    return this->warrior;
}

void Warlock::setWarriorID(WarriorID newID) {
    this->warrior = newID;
}

MagicianID Warlock::getMagicianID() const {
    return this->magician;
}

void Warlock::summonDemon() {
    this->getMagicPower().ensureHaveIsMana();
    this->returnDemon();
    
    if ( this->getMagicPower().getMana() < this->demonCost ) {
        std::cout << "Not enough mana!" << std::endl;
        return;
    }
    this->demon = new Demon();
        
    this->getMagicPower().takeMana(this->demonCost);
}

void Warlock::returnDemon() {
    if ( this->demon != nullptr ) {
        delete this->demon;
    }
}

File: Army(java)/Warriors/Warlock.java
package Warriors;

import EnumID.*;
import Basic.*;

public class Warlock extends SpellCaster {
    private int demonCost;
    private WarriorID warrior;
    private MagicianID magician;
    public Demon demon;

    public Warlock(String name,  int hp, int mana, int dmg) {
        super(name, hp, mana, dmg);
        this.demonCost = 50;
        this.warrior = WarriorID.WARLOCK;
        this.magician = MagicianID.SORCERER;
    }

    public WarriorID getWarriorID() {
        return warrior;
    }

    public void setWarriorID(WarriorID newID) {
        warrior = newID;
    }

    public MagicianID getMagicianID() {
        return magician;
    }

    public void summonDemon() throws MagicManaOffException  {
        this.getMagicPower().ensureHaveIsMana();
        
        if ( this.getMagicPower().getMana() < demonCost ) {
            System.out.println("Not enough mana!");
            return;
        }
        this.demon = new Demon("Demon", 150, 50);
            
        this.getMagicPower().takeMana(demonCost);
    }
}
